File: Data Structures & Algorithms/path-sum/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} targetSum
     * @return {boolean}
     */
    hasPathSum(root, targetSum) {
        const _helper = (node, sum=0) => {
            if (!node) return false;
            if (!node.left && !node.right) return sum === targetSum;
            const left = node.left ? node.left.val : 0;
            const right = node.right ? node.right.val : 0;
            return _helper(node.left, sum + left) || _helper(node.right, sum + right);
        }
        if (!root) return false;
        return _helper(root, root.val);
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and subRoot:
            return False
        if root and not subRoot:
            return True
        
        def is_same_tree(node1, node2):
            if not node1 and not node2:
                return True
            if (node1 and not node2) or (not node1 and node2):
                return False
            if node1.val != node2.val:
                return False
            
            return is_same_tree(node1.left, node2.left) and is_same_tree(node1.right, node2.right)
        
        if root.val == subRoot.val and is_same_tree(root, subRoot):
            return True

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
